File: DSA/LC/5_longest_palindromic_substring.py
# n^3 Brute Force
class Solution:
    def longestPalindrome(self, s: str) -> str:
        max_pal = ""
        combination = ""

        for i in range(len(s)):
            for j in range(len(s), i, -1):
                combination = s[i:j:]
                if self.isPalindrome(combination) and len(max_pal) < len(combination):
                    max_pal = combination

        return max_pal

    def isPalindrome(self, s: str) -> bool:
        start = 0
        end = len(s) - 1
        while start < end:
            if s[start] != s[end]:
                return False
            start += 1
            end -= 1
        return True

File: DSA/LC/5_longest_palindromic_substring.cpp
#include <string>

class Solution
{
public:
    std::string longestPalindrome(std::string s)
    {
        std::string max_pal{};
        int max_length{};

        for (int i = 0; i < s.size(); ++i)
        {
            int left = i;
            int right = i;
            // ODD
            while (left >= 0 && right < s.size() && s[left] == s[right])
            {
                if (right - left + 1 > max_length)
                {
                    max_pal = s.substr(left, right - left + 1);
                    max_length = right - left + 1;
                }
                --left;
                ++right;
            }

            left = i;
            right = i + 1;
            // EVEN
            while (left >= 0 && right < s.size() && s[left] == s[right])
            {
                if (right - left + 1 > max_length)
                {
                    max_pal = s.substr(left, right - left + 1);
                    max_length = right - left + 1;
                }
                --left;
                ++right;
            }
        }

        return max_pal;
    }
};
